packages/augur-state/src/db/SyncableDB.ts

```diff
@@ -1,100 +1,84 @@
+        fields: ['blockNumber']
+      }
+    });
+    dbController.notifySyncableDBAdded(this);
+  }
 
-        // TODO start blockstream
+  public async saveLogs(logs: Array<ParsedLog>, highestSyncedBlockNumber: number): Promise<void> {
+    let success = true;
+    if (logs.length > 0) {
+      const documents = _.sortBy(_.map(logs, this.processLog), "_id");
+      success = await this.bulkUpsertDocuments(documents[0]._id, documents);
     }
-
-    protected async getLogs(augur: Augur<TBigNumber>, startBlock: number, endBlock: number): Promise<Array<ParsedLog>> {
-        return await augur.events.getLogs(this.eventName, startBlock, endBlock);
+    if (success) {
+      await this.syncStatus.setHighestSyncBlock(this.dbName, highestSyncedBlockNumber);
     }
+  }
 
-    protected processLog(log: Log): BaseDocument {
-        if (!log.blockNumber) throw new Error(`Corrupt log: ${JSON.stringify(log)}`);
-        const _id = `${log.blockNumber.toPrecision(21)}${log.logIndex}`;
-        return Object.assign(
-            { _id },
-            log
-        );
+  public async addNewBlock(logs: Array<ParsedLog>): Promise<void> {
+    const highestSyncedBlockNumber = await this.syncStatus.getHighestSyncBlock(this.dbName);
+    const documents = _.sortBy(_.map(logs, this.processLog), "_id");
+    if (await this.bulkUpsertDocuments(documents[0]._id, documents)) {
+      await this.syncStatus.setHighestSyncBlock(this.dbName, highestSyncedBlockNumber + 1);
+    } else {
+      throw new Error(`Unable to add new block`);
     }
+  }
 
-    public async addNewBlock(logs: Array<ParsedLog>): Promise<void> {
-        const highestSyncedBlockNumber = await this.syncStatus.getHighestSyncBlock(this.dbName);
-        const documents = _.sortBy(_.map(logs, this.processLog), "_id");
-        if (await this.bulkUpsertDocuments(documents[0]._id, documents)) {
-            await this.syncStatus.setHighestSyncBlock(this.dbName, highestSyncedBlockNumber + 1);
-        } else {
-            throw new Error(`Unable to add new block`);
+  public async rollback(blockNumber: number): Promise<void> {
+    // Remove each change from blockNumber onward
+    try {
+      let highestSyncBlock = await this.syncStatus.getHighestSyncBlock(this.dbName);
+      // Sort blocks so newest blocks are removed first
+      let blocksToRemove = await this.db.find({
+        selector: { blockNumber: { $gte: blockNumber } },
+        fields: ['_id', 'blockNumber', '_rev'],
+        sort: [{blockNumber: 'desc'}],
+      });
+      if (blocksToRemove.docs.length > 0) {
+        console.log("\n\nDeleting the following blocks from " + this.dbName);
+        console.log(blocksToRemove.docs);
+        for (let doc of blocksToRemove.docs) {
+          // Remove block number from event DB
+          await this.db.remove(doc._id, doc._rev);
+          // Update highest sync block with decremented block number
+          await this.syncStatus.setHighestSyncBlock(this.dbName, --highestSyncBlock);
```

Comment:
Is blocksToRemove always going to have a block listed even if a certain block had no data? I'm unclear on what db.find returns with the selector and fields. This question is mainly for my own edification.

Reply:
Ah, good point. We should still be decrementing the highest sync block even if db.find doesn't return any results. There could certainly be cases where the query doesn't return any documents, but we still need to decrement (in case of a re-org).